Add documentation about PDBs (#769)
stackabletech/documentation: modules/concepts/pages/operations/pod_disruptions.adoc

Reviewer: Can we please rename the PR to something like `Add known issue on PDBs and certificate rotation`

Author: done

== Details
Have a look at the xref:contributor:adr/ADR030-allowed-pod-disruptions.adoc[ADR on Allowed Pod disruptions] for the implementation details.

> == Known issues with PDBs and e.g. certificate rotations

Reviewer: I *think* this is the only known issue, but I might be wrong
```suggestion
== Known issue with PDBs and certificate rotations
```

> PDBs together with certificate rotations can be problematic in case e.g. {commons-operator}[commons-operator] was unavailable until certificates expire. Assume a product like xref:zookeeper:index.adoc[Apache ZooKeeper] which needs a quorum and a PDB of `1`. If commons-operator comes back, it will rotate all expired certificates and thus `evict` pods. PDBs count exclusively for `eviction` and as the certificates are expired this pod will be stuck in `CrashLoopBackOff` forever unable to connect to the quorum, blocking PDB for the other pods. Other pods are unable to restart and rotate certificates.

Reviewer: note: For docs we normally have a newline after each sentence
```suggestion
PDBs together with certificate rotations can be problematic in case e.g. {commons-operator}[commons-operator] was unavailable to restart the Pods before the certificate expire.
commons-operator uses the `evict` API in Kubernetes, which respects the PDB.
If a Pod is evicted and a PDB would be violated, the Pod is *not* restarted.
Assume a product like xref:zookeeper:index.adoc[Apache ZooKeeper] which needs to form a quorum to function and the PDB only allows a single Pod to be unavailable.
As soon as enough certificates of the ZookeeperCluster have expired, all Pods will crash-loop, as they encounter expired certificates.
As only the container crash-loops (not the entire Pod), no new certificate is issued.
As soon as commons-operator comes online again it tries to `evict` a Zookeeper Pod.
However, this is prohibited, as the PDB would be violated.
```

NOTE: We encountered this problem only with the specific outlined case above and only under this circumstances.

=== Workaround

> If you encounter this only manually deleting those pods can help out of this situation as it is no `eviction`.

Reviewer: ```suggestion
If you encounter this only manually deleting those pods can help out of this situation.
A Pod deletion (other than evictions) does *not* respect PDBs, so the Pods can be restarted anyway.
All restarted Pods will get a new certificate, the stacklet should turn healthy again.
```

==== k9s
If you are using `k9s` you can start it in your terminal
[source, bash]
----
k9s
----
and type `0` to view all namespaces and then type e.g. `/zookeeper` and hit enter. Go with up and down to your pod and press `CTL + D` and confirm to delete the pod. Repeat with all other instances of the stuck product.

==== kubectl
List your pods with
[source, bash]
----
kubectl get pods -A
----
and copy the name of the pod you want to delete. Type
[source, bash]
----
kubectl delete pod zookeeper-server-default-0

> ----

Reviewer: I think users running into this are advanced enough to delete Pods.
I would simply mention this command and call it a day
```bash
kubectl delete pod -l app.kubernetes.io/name=zookeeper,app.kubernetes.io/instance=simple-zk
pod "simple-zk-server-default-0" deleted
pod "simple-zk-server-default-1" deleted
pod "simple-zk-server-default-2" deleted
```

Author: https://github.com/stackabletech/documentation/pull/769/commits/b81d587e16d38e5c258678aa1ff393b6a3dad256

to delete the instance with the name `zookeeper-server-default-0`. Repeat it for all instances of your product.

> === Preventing this Situation entirely

Reviewer: ```suggestion
=== Preventing this situation
```

> Only disabling PDBs can guarantee to not run into this scenario.

Reviewer: ```suggestion
The best measure is to make sure that commons-operator is always running, so that it can restart the Pods before the certificates expire.

A hacky way to prevent this situation could be to disable PDBs for the specific stacklet.
But this also has the downside, that you are now missing the benefits of the PDB.
```
